// n8n-nodes-base/dist/credentials/ImpervaWafApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ImpervaWafApi = void 0;
class ImpervaWafApi {
    constructor() {
        this.name = 'impervaWafApi';
        this.displayName = 'Imperva WAF API';
        this.documentationUrl = 'impervawaf';
        this.icon = 'file:icons/Imperva.svg';
        this.properties = [
            {
                displayName: 'API ID',
                name: 'apiID',
                type: 'string',
                default: '',
                required: true,
            },
            {
                displayName: 'API Key',
                name: 'apiKey',
                type: 'string',
                typeOptions: { password: true },
                default: '',
                required: true,
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    'x-API-Id': '={{$credentials.apiID}}',
                    'x-API-Key': '={{$credentials.apiKey}}',
                },
            },
        };
    }
}
exports.ImpervaWafApi = ImpervaWafApi;
