// n8n-nodes-base/dist/nodes/Airtable/v2/actions/record/search.operation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.execute = exports.description = void 0;
const utilities_1 = require("../../../../../utils/utilities");
const transport_1 = require("../../transport");
const utils_1 = require("../../helpers/utils");
const common_descriptions_1 = require("../common.descriptions");
const properties = [
    {
        displayName: 'Filter By Formula',
        name: 'filterByFormula',
        type: 'string',
        default: '',
        placeholder: "e.g. NOT({Name} = 'Admin')",
        hint: 'If empty, all the records will be returned',
        description: 'The formula will be evaluated for each record, and if the result is not 0, false, "", NaN, [], or #Error! the record will be included in the response. <a href="https://support.airtable.com/docs/formula-field-reference" target="_blank">More info</a>.',
    },
    {
        displayName: 'Return All',
        name: 'returnAll',
        type: 'boolean',
        default: true,
        description: 'Whether to return all results or only up to a given limit',
    },
    {
        displayName: 'Limit',
        name: 'limit',
        type: 'number',
        displayOptions: {
            show: {
                returnAll: [false],
            },
        },
        typeOptions: {
            minValue: 1,
            maxValue: 100,
        },
        default: 100,
        description: 'Max number of results to return',
    },
    {
        displayName: 'Options',
        name: 'options',
        type: 'collection',
        default: {},
        description: 'Additional options which decide which records should be returned',
        placeholder: 'Add Option',
        options: [
            {
                displayName: 'Download Attachments',
                name: 'downloadFields',
                type: 'multiOptions',
                typeOptions: {
                    loadOptionsMethod: 'getAttachmentColumns',
                    loadOptionsDependsOn: ['base.value', 'table.value'],
                },
                default: [],
                description: "The fields of type 'attachment' that should be downloaded",
            },
            {
                displayName: 'Output Fields',
                name: 'fields',
                type: 'multiOptions',
                typeOptions: {
                    loadOptionsMethod: 'getColumns',
                    loadOptionsDependsOn: ['base.value', 'table.value'],
                },
                default: [],
                description: 'The fields you want to include in the output',
            },
            common_descriptions_1.viewRLC,
        ],
    },
    {
        displayName: 'Sort',
        name: 'sort',
        placeholder: 'Add Sort Rule',
        description: 'Defines how the returned records should be ordered',
        type: 'fixedCollection',
        typeOptions: {
            multipleValues: true,
        },
        default: {},
        options: [
            {
                name: 'property',
                displayName: 'Property',
                values: [
                    {
                        displayName: 'Field',
                        name: 'field',
                        type: 'options',
                        typeOptions: {
                            loadOptionsMethod: 'getColumns',
                            loadOptionsDependsOn: ['base.value', 'table.value'],
                        },
                        default: '',
                        description: 'Name of the field to sort on. Choose from the list, or specify an ID using an <a href="https://docs.n8n.io/code-examples/expressions/">expression</a>.',
                    },
                    {
                        displayName: 'Direction',
                        name: 'direction',
                        type: 'options',
                        options: [
                            {
                                name: 'ASC',
                                value: 'asc',
                                description: 'Sort in ascending order (small -> large)',
                            },
                            {
                                name: 'DESC',
                                value: 'desc',
                                description: 'Sort in descending order (large -> small)',
                            },
                        ],
                        default: 'asc',
                        description: 'The sort direction',
                    },
                ],
            },
        ],
    },
];
const displayOptions = {
    show: {
        resource: ['record'],
        operation: ['search'],
    },
};
exports.description = (0, utilities_1.updateDisplayOptions)(displayOptions, properties);
async function execute(items, base, table) {
    let returnData = [];
    const body = {};
    const qs = {};
    const endpoint = `${base}/${table}`;
    try {
        const returnAll = this.getNodeParameter('returnAll', 0);
        const options = this.getNodeParameter('options', 0, {});
        const sort = this.getNodeParameter('sort', 0, {});
        const filterByFormula = this.getNodeParameter('filterByFormula', 0);
        if (filterByFormula) {
            qs.filterByFormula = filterByFormula;
        }
        if (options.fields) {
            if (typeof options.fields === 'string') {
                qs.fields = options.fields.split(',').map((field) => field.trim());
            }
            else {
                qs.fields = options.fields;
            }
        }
        if (sort.property) {
            qs.sort = sort.property;
        }
        if (options.view) {
            qs.view = options.view.value;
        }
        let responseData;
        if (returnAll) {
            responseData = await transport_1.apiRequestAllItems.call(this, 'GET', endpoint, body, qs);
        }
        else {
            qs.maxRecords = this.getNodeParameter('limit', 0);
            responseData = await transport_1.apiRequest.call(this, 'GET', endpoint, body, qs);
        }
        returnData = responseData.records;
        if (options.downloadFields) {
            return await transport_1.downloadRecordAttachments.call(this, responseData.records, options.downloadFields);
        }
        returnData = returnData.map((record) => ({
            json: (0, utils_1.flattenOutput)(record),
        }));
        const itemData = (0, utilities_1.generatePairedItemData)(items.length);
        returnData = this.helpers.constructExecutionMetaData(returnData, {
            itemData,
        });
    }
    catch (error) {
        if (this.continueOnFail()) {
            returnData.push({ json: { message: error.message, error } });
        }
        else {
            throw error;
        }
    }
    return returnData;
}
exports.execute = execute;
